driver/normalizer/strconv.go

```diff
@@ -29,15 +54,59 @@ func unquoteSingle(s string) (string, error) {
 	return string(buf), nil
 }
 
+// contains reports whether the string contains the byte c.
+func contains(s string, c byte) bool {
+	for i := 0; i < len(s); i++ {
+		if s[i] == c {
+			return true
+		}
+	}
+	return false
+}
+
+// replaceEscapedMaybe returns a copy of s with "\\old[^0-9]" replaced by new.
+func replaceEscapedMaybe(s string, old, new rune) string {
+	var runeTmp [utf8.UTFMax]byte
+	n := utf8.EncodeRune(runeTmp[:], new)
+
+	lastCp := 0
```

Comment:
Could this be simplified to something like the following (untested):

```go
i := 0
needle := "\\" + old
for i < len(s) {
   next := strings.Index(s[i:], needle)
   if next < 0 {
      break
   }
   // decode the first rune in s[i+next:] to verify it is not [0-9]
   i += next + runeSize
}
```

I think that would be slightly easier to follow than the very careful dance with indices.

Reply:
Separately, is this test actually correct? If a string contains, for example, `\u0000123` that is a valid encoding of [0, 49, 50, 51]. So I think checking for a non-digit after the escape may not be right: `\u` consumes 4 digits and `\U` consumes 8, unconditionally. 

And if this is really _only_ to handle the degenerate case of `\0`, can we specialize it to that instead of making a generic function that has to be so careful about lengths?

Reply:
> Separately, is this test actually correct?
I'm not sure if I understood the question right, but I have tried to add such test in fc24125ac5fe37c1e4c1c41321dbf6d9d01d3f58 and it passes with the current algorithm.

If that does not answer the question - could you please provide another possible input/output that would highlight the issue you see?

Reply:
On https://github.com/bblfsh/javascript-driver/pull/64#discussion_r264402621
> Could this be simplified to something like the following (untested):

I belive it can, although it's not a single-pass algorithm any more.

Not sure if that would be a great simplification - having 1 loop instead of 3 is indeed easier to follow, but one would still need a `buf` to copy memory and some state like `lastCp` that would track parts before the `needle` that are already transferred to `buf`, right?

Or do you suggest some other way to replace the `needle` with `new` rune ? Sorry if I misunderstood the suggestion somehow.

Reply:
Sorry for the confusion.

It is still single-pass, one pass over the input, assembling the output. I just think that the code here right now is overly micro-optimized for no clear need. Here is how I would simplify it:

```go
func replaceEscapedMaybe(s string, old, repl string) string {
	var out strings.Builder
	for s != "" {
		pos := strings.Index(s, old)
		if pos < 0 {
			break
		}
		out.WriteString(s[:pos])
		s = s[pos+len(old):]
		r, n := utf8.DecodeRuneInString(s)
		s = s[n:]
		if r >= '0' && r <= '9' {
			out.WriteString(old)
		} else {
			out.WriteString(repl)
		}
		if n != 0 {
			out.WriteRune(r)
		}

	}
	out.WriteString(s)
	return out.String()
}
```

There are two key points here: One is that we don't need any rune to string conversions, since this function is only ever called with fixed arguments so we can make them strings. The other is that instead of manually attempting to guess how the allocation should work, we're relying on a well-tested library component (the builder) to handle allocations for us. I did not benchmark this, but I would be very surprised if it's not at least as good. And the difference is not causing us a problem, as far as I can tell.

Reply:
Note too that this simplifies the logic for the degenerate case where there are no escapes: In that case, it fails the first index probe and returns the entire string. No more special case!

Reply:
Here are the updated benchmarks

```
goos: darwin
goarch: amd64
pkg: github.com/bblfsh/javascript-driver/driver/normalizer
BenchmarkReplacingNullEscape_Iterative-4   	10000000	       192 ns/op	      48 B/op	       2 allocs/op
BenchmarkReplacingNullEscape_Regexp-4      	 2000000	       891 ns/op	     120 B/op	       6 allocs/op
BenchmarkReplacingNullEscape_Simple-4      	10000000	       157 ns/op	      24 B/op	       2 allocs/op
PASS
ok  	github.com/bblfsh/javascript-driver/driver/normalizer	6.576s
```

It is not just simple, but even a bit faster! 👏

Reply:
Thank you very much for your kind explanation @creachadair, all points make perfect sense! 
Switched to your strings-based implementation in 8c08ccf and benchmark posted in https://github.com/bblfsh/javascript-driver/pull/64#issuecomment-472095705